proposals/low-level-struct-improvements.md

```diff
@@ -50,248 +45,226 @@ readonly ref struct Span<T>
+The first challenge is the span safety rules make a [hard assumption](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md#span-constructor) that `Span<T>` has no such constructor. By extension this also introduced a hard assumption that `ref` fields do not exist. This resulted in a significant simplification of the rules but it allows for a number of patterns that make introducing `ref` fields significantly more difficult. 
 
-```cs
-Span<int> CreatingAndReturningSpan()
-{
-    int i = 42;
-
-    // This must be an error in the new design because it stores stack 
-    // state in the Span. 
-    return new Span<int>(ref i);
+Consider the following method signature:
 
-    // This must be legal in the new design because it is legal today (it 
-    // cannot store stack state)
-    return new Span<int>(new int[] { });
+```c#
+Span<T> CreateSpan<T>(ref T parameter)
+{
+    // The implementation of this method is irrelevant when considering the lifetime of the 
+    // returned Span<T>. The rules disallow capture of `parameter` hence the return of 
+    // CreateSpan<T> is always safe-to-escape to the calling method
 }
 ```
 
-At the same time it is legal to have methods today which take a `ref` parameter and return a `Span<T>`.  These methods bear a lot of similarity to the newly added `Span<T>` constructor: take a `ref`, return a `Span<T>`. However the lifetime of the return value of these methods is never restricted by the inputs.  The existing span safety rules consider such values as effectively always *safe-to-escape* outside the enclosing method.
+By the existing span safety rules the return of an invocation of this method always has a //*safe-to-escape* value of *calling method*. That is because in the span safety document `ref` fields do not exist and hence there is no way for the return to capture `parameter` hence there is no constraint on the return.  The implementation of the method is completely irrelevant here, which is why it's omitted from the sample, because there is simply no way for the `ref` to be captured. This means no matter how this method is called the return is *safe-to-escape* to the *calling method*.
 
-```cs
-class ExistingScenarios
+```c#
+Span<T> CompatExample(ref int p)
 {
-    Span<T> CreateSpan<T>(ref T p)
-    {
-        // The implementation of this method is irrelevant. From the point of
-        // the consumer the returned value is always safe to return.
-        ... 
-    }
-
-    Span<T> Examples<T>(ref T p, T[] array)
-    {
-        // Legal today
-        return CreateSpan(ref p); 
+    // Okay
+    int local = 42;
+    return CreateSpan<int>(ref local);
 
-        // Legal today, must remain legal
-        T local = default;
-        return CreateSpan(ref local);
+    // Okay
+    Span<T> span = stackalloc int[1];
+    return CreateSpan<int>(ref span[0]);
```

Comment:
wait... why is this ok? :)   furthermore... if it was legal before... woudl it be bad to become illegal?

Reply:
This was legal because there was no mechanism by which `CreateSpan(ref  int local)` could every capture and return the `ref`. Basically there is no way given the rules of C# and the safe APIs of dotnet/runtime to make the returned `Span<T>` refer to the passed in value. 

Yes you could do this using `MemoryMarshal` or `Unsafe` but those are APIs that are considered `unsafe`. They are essentially outside the bounds of span safety guidelines (and are documented as such)

Reply:
@jaredpar, I think the "issue" here is that the code is confusing to read.

If you were to think about this from the perspective of the framework design guidelines and how the BCL names things; the "logical assumption" of `CreateSpan<int>(ref T)` is that it creates a `span` over the `ref` input and so the readers initial "assumption" is that `CreateSpan<int>(stackalloc)` is returning a span that captures the stack allocation.

The language prevents this today (outside of unsafe code), but its still not how the code reads. This is somewhat made more "confusing" by the fact that part of this doc is meant to enable scenarios where `CreateSpan(ref T)` capturing the input is possible.

Perhaps it would be more clear if this was `CreateSpanThatDoesntCaptureTheInput` (e.g. a method without `RefThisEscapes`) and `CreateSpanThatCapturesTheInput` (e.g. a method with `RefThisEscapes`)?
-- or better names, because names are hard and my suggestions are terrible 😄

Reply:
The goal of this section though is to really drive home the point that 

1. This method creates a `Span<T>` 
2. It can never, under any circumstances, capture that input value

If I name it say `CreateSpanThatDoesntCaptureTheInput` it might imply that this particular method does not capture the input but maybe some other method could. Where the problem is there is just no way this can happen, no matter the name or implementatoin.

Reply:
> If I name it say CreateSpanThatDoesntCaptureTheInput it might imply that this particular method does not capture the input but maybe some other method could. Where the problem is there is just no way this can happen, no matter the name or implementatoin.

I thought that was the point of `RefThisEscapes`? `RefThisEscapes` means that you can have a span that captures the `ref input`. The language would then block it in the case of `stackalloc` because its unsafe.

That is, there are at least four basic scenarios to consider here, three of which are legal:
```csharp
// this is legal as `createdSpan` can't capture `ref span[0]` as it isn't attributed with `RefThisEscapes`
Span<T> stackSpan = stackalloc int[1];
var createdSpan = CreateSpanWithoutRefThisEscapes<int>(ref span[0]);

// this is legal as `createdSpan` can't capture `ref span[0]` as it isn't attributed with `RefThisEscapes`
Span<T> stackSpan = stackalloc int[1];
return CreateSpanWithoutRefThisEscapes<int>(ref span[0]);

// this is legal as even though `createdSpan` can capture `ref span[0]` since it is attributed with `RefThisEscapes`, it isn't escaping the scope of `stackSpan`
Span<T> stackSpan = stackalloc int[1];
var createdSpan = CreateSpanWithRefThisEscapes<int>(ref span[0]);

// this is ILLEGAL as `createdSpan` can capture `ref span[0]` and if it did, it would allow `stackSpan` to escape its scope
Span<T> stackSpan = stackalloc int[1];
return CreateSpanWithRefThisEscapes<int>(ref span[0]);
```

Reply:
> RefThisEscapes means that you can have a span that captures the ref input

No. The only thing that `RefThisEscapes` alters is that it makes it possible to return `this` by-ref from instance methods on `struct`. It does not change your ability to capture a `ref` as a field.

Reply:
Maybe I'm getting the attributes mixed up then?

This allows ref fields, which means that you can functionally `new Span(ref param)` and the created `Span` will have the scope of the passed input.

So if the `CreateSpan` call uses that method and returns the span, its lifetime is the same as the input and the compiler does the right analysis.